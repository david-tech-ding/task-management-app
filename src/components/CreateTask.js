import React, { useState } from "react";
import axios from "axios";
import DatePicker from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";
import "../styles/createtask.css";

const CreateTask = () => {
  axios.defaults.baseURL = "https://api.trello.com/1";

  const [date, setDate] = useState();

  const initialState = {
    fields: {
      title: "",
      details: "",
      priorityLevel: "High",
      assignTo: "",
      dueDate: "",
    },
  };
  const [fields, setFields] = useState(initialState.fields);
  const handleCreateTask = (e) => {
    e.preventDefault();
    console.log(fields);
    axios
      .post("/cards", { ...fields })
      .then(() => {
        console.log(fields);
      })
      .catch((err) => console.log(err));
  };
  const handleDateChange = (chosenDate) => {
    setDate(chosenDate);
    setFields({ ...fields, dueDate: chosenDate });
  };
  const handleFieldChange = (e) => {
    e.preventDefault();
    setFields({ ...fields, [e.target.name]: e.target.value });
  };

  return (
    <div className="create-task">
      <h2 className="create-task-heading">Create A New Task</h2>
      <form className="create-task-form" onSubmit={handleCreateTask}>
        <label htmlFor="title">
          Title
          <input
            className="create-task-form_input"
            type="text"
            id="title"
            name="title"
            value={fields.title}
            onChange={handleFieldChange}
          />
        </label>
        <label htmlFor="details">
          Details
          <input
            className="create-task-form_input"
            type="text"
            id="details"
            name="details"
            value={fields.details}
            onChange={handleFieldChange}
          />
        </label>
        <label htmlFor="priorityLevel">
          Priority Level
          <select
            className="create-task-form_select"
            id="priorityLevel"
            name="priorityLevel"
            value={fields.priorityLevel}
            onChange={handleFieldChange}
          >
            <option className="priority-level-high" value="High">
              High
            </option>
            <option className="priority-level-high" value="Medium">
              Medium
            </option>
            <option className="priority-level-high" value="Low">
              Low
            </option>
          </select>
        </label>
        <label htmlFor="assignTo">
          Assign To
          <input
            className="create-task-form_input"
            type="text"
            id="assignTo"
            name="assignTo"
            value={fields.assignTo}
            onChange={handleFieldChange}
          />
        </label>
        <label htmlFor="dueDate">
          Due Date
          <DatePicker
            className="create-task-form_input"
            selected={date}
            onChange={handleDateChange}
            id="dueDate"
            name="dueDate"
            value={fields.dueDate}
          />
        </label>
        <button type="submit">Create</button>
      </form>
    </div>
  );
};

export default CreateTask;
