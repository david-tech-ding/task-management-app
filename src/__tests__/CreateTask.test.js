import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import axios from "axios";
import { MemoryRouter as Router } from "react-router-dom";
import CreateTask from "../components/CreateTask";

jest
  .spyOn(axios, "post")
  .mockImplementation(() => Promise.resolve({ data: {} }));

describe("CreateTask component", () => {
  it("renders correctly", () => {
    const users = [
      { id: 1, firstName: "David", lastName: "Ding", jobRole: "Developer" },
      {
        id: 2,
        firstName: "James",
        lastName: "Lee",
        jobRole: "Product Manager",
      },
      {
        id: 3,
        firstName: "Jasper",
        lastName: "Lim",
        jobRole: "Human Resources",
      },
    ];

    const user = {
      id: 3,
      firstName: "Jasper",
      lastName: "Lim",
      jobRole: "Human Resources",
    };

    const { asFragment } = render(
      <Router>
        <CreateTask user={user} users={users} />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  //   it("creates a new Task when the form is submitted", async () => {
  //     const users = [
  //       { id: 1, firstName: "David", lastName: "Ding", jobRole: "Developer" },
  //       {
  //         id: 2,
  //         firstName: "James",
  //         lastName: "Lee",
  //         jobRole: "Product Manager",
  //       },
  //       {
  //         id: 3,
  //         firstName: "Jasper",
  //         lastName: "Lim",
  //         jobRole: "Human Resources",
  //       },
  //     ];

  //     const user = {
  //       id: 3,
  //       firstName: "Jasper",
  //       lastName: "Lim",
  //       jobRole: "Human Resources",
  //     };

  //     render(
  //       <Router>
  //         <CreateTask user={user} users={users} />
  //       </Router>
  //     );

  //     const titleInput = screen.getByLabelText("Title");
  //     const detailsInput = screen.getByLabelText("Details");
  //     const priorityDropdown = screen.getByTestId("priorityLevel");
  //     const assignToDropdown = screen.getByTestId("assignTo");
  //     const dueDateInput = screen.getByLabelText("Due Date");
  //     const createButton = screen.getByText("Create");

  //     fireEvent.change(titleInput, { target: { value: "Test Title" } });
  //     fireEvent.change(detailsInput, { target: { value: "Test Details" } });
  //     fireEvent.change(priorityDropdown, { target: { value: "High" } });
  //     fireEvent.change(assignToDropdown, { target: { value: "1" } });
  //     fireEvent.click(dueDateInput);
  //     fireEvent.click(screen.getByText("8"));

  //     fireEvent.click(createButton);

  //     await waitFor(() => {
  //       expect(axios.post).toHaveBeenCalledWith("/task", {
  //         title: "Test Title",
  //         details: "Test Details",
  //         priorityLevel: "High",
  //         assignTo: "1",
  //         dueDate: "08/08/2023",
  //       });
  //     });

  //     expect(titleInput.value).toBe("Test Title");
  //     expect(detailsInput.value).toBe("Test Details");
  //     expect(priorityDropdown.value).toBe("High");
  //     expect(assignToDropdown.value).toBe("1");
  //     expect(dueDateInput.value).toBe("08/08/2023");
  //   });
});
