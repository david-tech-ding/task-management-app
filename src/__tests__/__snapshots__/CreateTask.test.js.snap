// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTask component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="create-task"
  >
    <h2
      class="create-task-heading"
    >
      Create A New Task
    </h2>
    <form
      class="create-task-form"
    >
      <label
        for="title"
      >
        Title
        <input
          class="create-task-form_input"
          id="title"
          name="title"
          type="text"
          value=""
        />
      </label>
      <label
        for="details"
      >
        Details
        <input
          class="create-task-form_input"
          id="details"
          name="details"
          type="text"
          value=""
        />
      </label>
      <label
        for="priorityLevel"
      >
        Priority Level
        <select
          class="create-task-form_select"
          data-testid="priorityLevel"
          id="priorityLevel"
          name="priorityLevel"
        >
          <option
            class="priority-level-high"
            value="High"
          >
            High
          </option>
          <option
            class="priority-level-high"
            value="Medium"
          >
            Medium
          </option>
          <option
            class="priority-level-high"
            value="Low"
          >
            Low
          </option>
        </select>
      </label>
      <label
        for="assignTo"
      >
        Assign To
        <select
          class="create-task-form_input"
          data-testid="assignTo"
          id="assignTo"
          name="assignTo"
        >
          <option
            value="1"
          >
            David Ding (Developer)
          </option>
          <option
            value="2"
          >
            James Lee (Product Manager)
          </option>
          <option
            value="3"
          >
            Jasper Lim (Human Resources)
          </option>
        </select>
      </label>
      <label
        for="dueDate"
      >
        Due Date 
      </label>
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container"
        >
          <input
            autocomplete="off"
            class="create-task-form_input"
            id="dueDate"
            name="dueDate"
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        type="submit"
      >
        Create
      </button>
    </form>
  </div>
</DocumentFragment>
`;
